src/routes/projects.rs

```diff
@@ -1085,102 +1113,432 @@ pub async fn project_edit(
                 .await?;
             }
 
-            if let Some(project) = &new_project.flame_anvil_project {
-                if !perms.contains(Permissions::EDIT_DETAILS) {
+            transaction.commit().await?;
+            Ok(HttpResponse::NoContent().body(""))
+        } else {
+            Err(ApiError::CustomAuthentication(
+                "You do not have permission to edit this project!".to_string(),
+            ))
+        }
+    } else {
+        Ok(HttpResponse::NotFound().body(""))
+    }
+}
+
+#[derive(Deserialize, Validate)]
+pub struct BulkEditProject {
+    #[validate(length(max = 3))]
+    pub categories: Option<Vec<String>>,
+    #[validate(length(max = 3))]
+    pub add_categories: Option<Vec<String>>,
+    pub remove_categories: Option<Vec<String>>,
+
+    #[validate(length(max = 256))]
+    pub additional_categories: Option<Vec<String>>,
+    #[validate(length(max = 3))]
+    pub add_additional_categories: Option<Vec<String>>,
+    pub remove_additional_categories: Option<Vec<String>>,
+
+    #[validate]
+    pub donation_urls: Option<Vec<DonationLink>>,
+    #[validate]
+    pub add_donation_urls: Option<Vec<DonationLink>>,
+    #[validate]
+    pub remove_donation_urls: Option<Vec<DonationLink>>,
+
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "::serde_with::rust::double_option"
+    )]
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 2048)
+    )]
+    pub issues_url: Option<Option<String>>,
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "::serde_with::rust::double_option"
+    )]
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 2048)
+    )]
+    pub source_url: Option<Option<String>>,
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "::serde_with::rust::double_option"
+    )]
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 2048)
+    )]
+    pub wiki_url: Option<Option<String>>,
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "::serde_with::rust::double_option"
+    )]
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 2048)
+    )]
+    pub license_url: Option<Option<String>>,
```

Comment:
License URL shouldn't be bulk editable, authors shouldn't be able to blindly set a license URL without having right next to it the license they have set

Reply:
oops, true